### DISCUSSION TITLE
easy way to replace a button tooltip

### DISCUSSION THREAD

**Discussion by klauswiesel** [stephanrauh/ngx-extended-pdf-viewer#1124] [Q&A]
Is there an easier way to replace the tooltip of a button than to include the complete json in the index.html? We are working on a project with 10+ languages supported...

**Comment by stephanrauh** [comment]
What are you trying to do? If I get you right, you

- need the i18n feature
- you want to be able to modify some of the default translations,
- and you've modified the json snippets provided by the library, but that's cumbersome.

If that's correct, I recommend using the raw translation files. Adding the json files to the index.html is a good idea for one or two languages, but if you need more languages, the default strategy is faster.

My default recipe https://pdfviewer.net/extended-pdf-viewer/getting-started instructs you to copy the entire `assets` folder. This includes the language files. I suppose you can overwrite some of these files in a second step by configuring the `angular.json`. Something like that:

```json
"assets": [
    "src/favicon.ico",
    "src/assets",
    {
      "glob": "**/*",
      "input": "node_modules/ngx-extended-pdf-viewer/assets/",
      "output": "/assets/"
    },
    {
      "glob": "**/*",
      "input": "custom-translations/locale",
      "output": "/assets/locale/"
    }
  ],
```

The idea being you take one of the translation file (e.g. [this one](https://github.com/stephanrauh/ngx-extended-pdf-viewer/blob/main/projects/ngx-extended-pdf-viewer/assets/locale/de/viewer.properties)), modify it, and store the modified version in the folder `custom-translations/locale/de/viewer.properties`.

Does this help you?

Best regards,
Stephan

**Comment by klauswiesel** [comment]
Hi Stephan

The way you are showing seems ok, it looks just a little cumbersome, as it enforces to check each new version on our side if labels are changed on your side.
I was just thinking if there is something that reads your defaults and allows to overwrite the label at runtime.

Thanks for your support
Klaus

**Reply by stephanrauh** [reply]
I see. Maybe you can help me because I'm very busy this week. The source code that needs to be modified is here: https://github.com/stephanrauh/pdf.js/tree/bleeding-edge/external/webL10n

More to the point, I guess you'll want to modify this method: https://github.com/stephanrauh/pdf.js/blob/de1d458022c0095321eeb274753181c632f7693c/external/webL10n/l10n.js#L327. You could load the default i18n bundle and merge it with a custom i18n bundle somewhere in this method.

BTW, by the look of it, you're working for an interesting company! :)

Best regards,
Stephan

**Comment by klauswiesel** [comment]
I used the approach you proposed regarding complete files…

**Reply by stephanrauh** [reply]
... and you're happy with this approach? If you aren't, feel free to open an issue. It sounds like an exotic use-case, but maybe it's not that far-fetched: I've added a couple of buttons and menu items without being able to provide translations for a 100+ languages. Currently, developers have a hard time translating the labels of these features to their local language.
